metadata_layer.py

```diff
@@ -351,36 +351,188 @@ def __repr__(self):
         self.licenses_owned = licenses_owned
         self.licenses_available = licenses_available
         self.licenses_reserved = licenses_reserved
         self.patrons_in_hold_queue = patrons_in_hold_queue
+
+        # format contains pdf/epub, drm, link
+        self.formats = formats or []
+
+        self.default_rights_uri = None
+        self.set_default_rights_uri(data_source=self.data_source, default_rights_uri=default_rights_uri)
         self.first_appearance = first_appearance
         self.last_checked = last_checked or datetime.datetime.utcnow()
 
+        self.__links = None
+        self.links = links
+        
+    @property
+    def links(self):
+        return self.__links
+
+    @links.setter
+    def links(self, arg_links):
+        """ If got passed all links, undiscriminately, filter out to only those relevant to pools. """
+        # start by deleting any old links
+        self.__links = []
+
+        if not arg_links:
+            return
+
+        for link in arg_links:
+            if link.rel in [Hyperlink.OPEN_ACCESS_DOWNLOAD, Hyperlink.DRM_ENCRYPTED_DOWNLOAD]:
+                # only accept the types of links relevant to pools
+                self.__links.append(link)
+
+                # If a link has a rights_uri, make that the overall rights_uri. 
+                # If there are multiple links with a rights_uri, they should have gotten 
+                # split into separate metadata objects in the content server or whatever code called this. 
+                if link.rights_uri:
+                    self.set_default_rights_uri(data_source=self.data_source, default_rights_uri=link.rights_uri)
+
+                # An open-access link or open-access rights implies a FormatData object.
+                open_access_link = (link.rel == Hyperlink.OPEN_ACCESS_DOWNLOAD
+                                    and link.href)
+                
+                open_access_rights_link = (link.media_type in Representation.BOOK_MEDIA_TYPES 
+                                           and link.href
+                                           and self.default_rights_uri in RightsStatus.OPEN_ACCESS)
+                
+                #if open_access_link or open_access_rights_link:
+                if open_access_link:
+                    self.formats.append(
+                        FormatData(
+                            content_type=link.media_type,
+                            drm_scheme=DeliveryMechanism.NO_DRM,
+                            link=link
+                        )
+                )
+
+
+    def __repr__(self):
+        description_string = '<CirculationData primary_identifier="%r" licenses_owned="%s" licenses_available="%s" licenses_reserved=%s' % (
+            self.primary_identifier, self.licenses_owned, self.licenses_available, self.licenses_reserved
+        )
+        description_string += 'patrons_in_hold_queue="%s" default_rights_uri="%s" first_appearance="%s" last_checked="%s" >' % (
```

Comment:
This is kind of a mess. I recommend you define the entire template at once:

'<CirculationData primary_identifier="%(primary_identifier)r" licenses_owned="%(licenses_owned)d" licenses_available="%(licesnses_available)d" licenses_reserved=%(licenses_reserved)d patrons_in_hold_queue="%(patrons_in_hold_queue)d" default_rights_uri="%(default_rights_uri)s" first_appearance="%(first_appearance)s" last_checked="%(last_checked)s" links="%(links)r" formats="%(formats)r" data_source="%(data_source_name)s">'

Then you can create a dictionary with the values for each interpolation variable, and then format the whole thing at once:

`return description_template % variables`